server/route.go

```diff
@@ -1396,98 +1616,311 @@ func (s *Server) createRoute(conn net.Conn, rURL *url.URL) *client {
+		c.route.poolIdx = idx
+		rtype := c.route.routeType
 		cid := c.cid
-		hash := c.route.hash
 		idHash := c.route.idHash
+		rHash := c.route.hash
+		rn := c.route.remoteName
+		url := c.route.url
+		// For solicited routes, we need now to send the INFO protocol
+		if didSolicit {
+			c.enqueueProto(s.generateRouteInitialInfoJSON(_EMPTY_, idx))
+		}
 		c.mu.Unlock()
 
+		// Add to the slice and bump the count of connections for this remote
+		conns[idx] = c
+		sz++
+		// This boolean will indicate that we are registering the only
+		// connection in non pooled situation or we stored the very first
+		// connection for a given remote server.
+		doOnce := !pool || sz == 1
+		if doOnce {
+			// check to be consistent and future proof. but will be same domain
+			if s.sameDomain(info.Domain) {
+				s.nodeToInfo.Store(rHash,
+					nodeInfo{rn, s.info.Version, s.info.Cluster, info.Domain, id, nil, nil, nil, false, info.JetStream})
+			}
+		}
+
 		// Store this route using the hash as the key
-		s.storeRouteByHash(hash, idHash, c)
+		if pool {
+			idHash += strconv.Itoa(idx)
+		}
+		s.storeRouteByHash(idHash, c)
 
 		// Now that we have registered the route, we can remove from the temp map.
 		s.removeFromTempClients(cid)
 
-		// we don't need to send if the only route is the one we just accepted.
-		sendInfo = len(s.routes) > 1
+		if doOnce {
+			// If the INFO contains a Gateway URL, add it to the list for our cluster.
+			if info.GatewayURL != _EMPTY_ && s.addGatewayURL(info.GatewayURL) {
+				s.sendAsyncGatewayInfo()
+			}
+
+			// we don't need to send if the only route is the one we just accepted.
+			if len(s.routes) > 1 {
+				// Now let the known servers know about this new route
+				s.forwardNewRouteInfoToKnownServers(info)
+			}
+
+			// Send info about the known gateways to this route.
+			s.sendGatewayConfigsToRoute(c)
+
+			// Unless disabled, possibly update the server's INFO protocol
+			// and send to clients that know how to handle async INFOs.
+			if !opts.Cluster.NoAdvertise {
+				s.addConnectURLsAndSendINFOToClients(info.ClientConnectURLs, info.WSConnectURLs)
+			}
+
+			// Add the remote's leafnodeURL to our list of URLs and send the update
+			// to all LN connections. (Note that when coming from a route, LeafNodeURLs
+			// is an array of size 1 max).
+			if len(info.LeafNodeURLs) == 1 && s.addLeafNodeURL(info.LeafNodeURLs[0]) {
+				s.sendAsyncLeafNodeInfo()
+			}
+		}
+
+		// Send the subscriptions interest.
+		s.sendSubsToRoute(c, idx, _EMPTY_)
 
-		// If the INFO contains a Gateway URL, add it to the list for our cluster.
-		if info.GatewayURL != "" && s.addGatewayURL(info.GatewayURL) {
-			s.sendAsyncGatewayInfo()
+		// In pool mode, if we did not yet reach the cap, try to connect a new connection
+		if pool && didSolicit && sz != s.routesPoolSize {
+			s.startGoRoutine(func() {
+				select {
+				case <-time.After(time.Duration(rand.Intn(100)) * time.Millisecond):
+				case <-s.quitCh:
+					s.grWG.Done()
```

Comment:
Just defer at top of func for optics? Behavior is correct already I think.

Reply:
Actually no (and I have added a comment in the upcoming changes). The function connectToRoute() is itself calling s.grWG.Done(), so we need to do it only if that function is not invoked. Otherwise it would lead to two calls to Done() for the same Add(), which is bad.